plugins/notion/src/api.ts

```diff
@@ -0,0 +1,464 @@
+    })
+}
+
+export async function getNotionDatabases() {
+    if (!notion) {
+        initNotionClient()
+    }
+
+    const results = await collectPaginatedAPI(notion!.search, {
+        filter: {
+            property: "object",
+            value: "database",
+        },
+    })
+
+    return results.filter(isFullDatabase)
+}
+
+export function assertFieldTypeMatchesPropertyType(
+    propertyType: NotionProperty["type"],
+    fieldType: ManagedCollectionField["type"]
+): asserts fieldType is ManagedCollectionField["type"] {
+    if (!isSupportedPropertyType(propertyType)) {
+        throw new Error(`Property type '${propertyType}' is not supported.`)
+    }
+
+    const allowedFieldTypes = supportedCMSTypeByNotionPropertyType[propertyType]
+
+    if (!allowedFieldTypes.includes(fieldType as never)) {
+        throw new Error(`Field type '${fieldType}' is not valid for property type '${propertyType}'.`)
+    }
+}
+
+export async function getDatabase(databaseId: string) {
+    if (!notion) {
+        initNotionClient()
+    }
+
+    assert(notion, "Notion client is not initialized")
+    const database = await notion.databases.retrieve({ database_id: databaseId })
+
+    if (!isFullDatabase(database)) {
+        throw new Error(`Database ${databaseId} is not a full database`)
+    }
+
+    return database
+}
+
+export function richTextToPlainText(richText: RichTextItemResponse[] | undefined) {
+    return Array.isArray(richText) ? richText.map(value => value.plain_text).join("") : ""
+}
+
+function isSupportedPropertyType(type: string): type is keyof typeof supportedCMSTypeByNotionPropertyType {
+    return type in supportedCMSTypeByNotionPropertyType
+}
+
+export function getDatabaseFieldsInfo(database: GetDatabaseResponse, databaseIdMap: DatabaseIdMap) {
+    const result: FieldInfo[] = []
+
+    // This property is always there but not included in `"database.properties"
+    result.push(pageContentProperty)
+
+    const supported: FieldInfo[] = []
+    const unsupported: FieldInfo[] = []
+    const missingCollection: FieldInfo[] = []
+
+    for (const key in database.properties) {
+        const property = database.properties[key]
+        assert(property)
+
+        const allowedTypes = isSupportedPropertyType(property.type)
+            ? supportedCMSTypeByNotionPropertyType[property.type]
+            : []
+        const fieldInfo: FieldInfo = {
+            id: property.id,
+            name: property.name,
+            originalName: property.name,
+            type: allowedTypes[0] ?? null,
+            allowedTypes,
+            notionProperty: property,
+        }
+
+        if (isMissingCollection(fieldInfo, databaseIdMap)) {
+            missingCollection.push(fieldInfo)
+            continue
+        }
+
+        const isUnsupported = !Array.isArray(allowedTypes) || allowedTypes.length === 0
+        if (isUnsupported) {
+            unsupported.push(fieldInfo)
+        } else {
+            supported.push(fieldInfo)
+        }
```

Comment:
Nit: Flip the logic maybe? The fewer negations, the easier to read usually (no "un" in "unsupported").